### PR TITLE
meta: added chromatic for visual regression tests

### PR DESCRIPTION
**Pull request by ovflowd**
## Description

This PR removes Storybook Test Runner (as it is pretty unstable and was a major headache for us) and replaces it with managed Visual Regression Tests, Provided by a platform named Chromatic (that gently offered us an Open Source plan).

This PR removes all Storybook Test Runner mentions, The extra Snapshots and updates our GitHub Action steps with the Visual Regression Tests.

The Snapshot tests weren't really useful because the actual motivation of the Snapshots is to detect changes. But that became increasingly more challenging with class names, styles, or URLs. Visual Regression Tests should easily detect these invisible or visible changes.

This also removes a few dependencies from us to maintain and removes extra complexity from our GitHub Actions.

## Validation

This PR should introduce Chromatic as a Pull Request step; It should run against this PR and pass.

### CODE REVIEW

**Comment by ovflowd** [comment]
cc @nodejs/website as I'd pretty much would like some reviews here :)

**Comment by ovflowd** [comment]
**Note.** With the usage of Chromatic, building our Stories directly on Vercel isn't necessary anymore, as Chromatic also deploys Storybook's on their own CDN.

It allows us to focus our Vercel infrastructure on what matters.

cc @nodejs/website, as I'd love to hear y'all opinion.

**Comment by AugustinMauroy** [comment]
One point for snapshot:
When I review on pr I start by taking a look to snapshot. With that we saw html structure of component.

Also we used snapshot on nodejs.dev to unsure component render correctly. And now we use storybook, so why not remove it.

**Comment by ovflowd** [comment]
> When I review on pr I start by taking a look to snapshot. With that we saw html structure of component.

You can see the structure of a Component on the Component itself. The Snapshot (in the current infrastructure) is just a string of the resulting HTML. It doesn't bring any value; I as a reviewer have 0 understanding of how the Component works by looking at a HTML snapshot.

It's also important to mention that these snapshots are flaky as they do not necessarily wait for the rendering or re-rendering of the Component, which was a big headache for a long time.

> Also we used snapshot on nodejs.dev to unsure component render correctly. And now we use storybook, so why not remove it.

That's not indicative of anything. We used Snapshots there because that was the infrastructure set up there. Snapshots do not indicate a Component rendering correctly, it's just a string. A Visual Regression Test is indicative of a component rendering correctly.

**Comment by ovflowd** [comment]
Also, @AugustinMauroy, those are just my views. But I'd love to understand your workflow better, and I'm fine to keep Snapshots if the team believes there's value here.

Hence, we should keep this PR open for a while, to hear the @nodejs/website team's opinions.

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -159,10 +159,13 @@ jobs:
...
-        # We want to enforce that the actual `turbo@latest` package is used instead of a possible hijack from the user
-        # the `${{ needs.base.outputs.turbo_args }}` is a string substitution happening from the base job
-        run: npx --package=turbo@latest -- turbo test:storybook:local ${{ needs.base.outputs.turbo_args }} -- --ci
+      - name: Run Storybook Visual Regression Tests
+        # We use Chromatic for Visual Regression Testing
+        # Note that this token is not a secret and safe to be added directly on the CI
+        # This Step wll fail if there are Visual Regressions and then should be acted on on Chromatic's UI or Review Comments
+        # See https://www.chromatic.com/docs/review
+        # If you're a Node.js Website Collaborator and do not have access to Chromatic, please request access to the Website Team Maintainers
+        run: npx --package=chromatic@latest -- chromatic --project-token=chpt_17c2d0dc8a68711 --build-script-name=storybook:build --only-changed --ci
```
Comment (araujogui) [inline .github/workflows/pull-request.yml]: Why the not the [chromatic action](https://github.com/marketplace/actions/publish-to-chromatic)?

File: package.json [hunk 2]
```diff
@@ -99,6 +98,8 @@
     "handlebars": "4.7.7",
     "jest": "29.6.1",
     "jest-environment-jsdom": "29.6.1",
+    "jest-junit": "16.0.0",
```
Comment (araujogui) [inline package.json:100]: What are these packages for

**Review by araujogui — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (ovflowd) [reply package.json:100]: Which packages? jest-junit is used by the junit reporter. It was being provided by @storybook/test-runner as an implicit dependency, but now I added as an explicit one.

File: .github/workflows/pull-request.yml (hunk 1 above)
Comment (ovflowd) [reply .github/workflows/pull-request.yml]: No real need to be honest. In the end it does the same thing, or do you see any benefit over invoking the package? (I’m fine either way)

**Comment by ovflowd** [comment]
cc @bmuenzenmeyer and @araujogui for reviews

File: package.json [hunk 3]
```diff
@@ -31,11 +36,7 @@
     "storybook:build": "cross-env NODE_NO_WARNINGS=1 storybook build --quiet",
     "test:unit": "cross-env NODE_NO_WARNINGS=1 jest --passWithNoTests",
     "test:unit:watch": "npm run test:unit -- --watch",
-    "test:storybook": "test-storybook --no-index-json",
-    "test:storybook:local": "concurrently -P -r -k -s first npm:storybook \"wait-on http://localhost:6006 && npm run test:storybook -- {@}\"",
-    "test:storybook:snapshot": "npm run test:storybook:local -- -- --updateSnapshot",
-    "test:storybook:watch": "npm run test:storybook:local -- -- --watch",
-    "test": "concurrently -r -s all npm:test:unit npm:test:storybook:local",
+    "test": "npm run test:unit",
```
Comment (AugustinMauroy) [inline package.json:39]: why we nee that simple test:unit work fine ?

Comment (ovflowd) [reply package.json:39]: I didn't get your question.

Comment (AugustinMauroy) [reply package.json:39]: why we add this? Why we don't simple keep `test:unit` ?

Comment (ovflowd) [reply package.json:39]: It's common practice for app projects to have a "test" and "build" commands. This reduces the learning gap if someone wants to test our repository easily.

**Review by araujogui — APPROVED** [review]

File: package.json (hunk 3 above)
Comment (AugustinMauroy) [reply package.json:39]: Ok that make sens. +1 to keep that

**Review by bmuenzenmeyer — APPROVED** [review]
> Provided by a platform named Chromatic

for context to others, Chromatic maintains Storybook, so this is a really nice fit

> The Snapshot tests weren't really useful because the actual motivation of the Snapshots is to detect changes. But that became increasingly more challenging with class names, styles, or URLs. Visual Regression Tests should easily detect these invisible or visible changes.

Big +1 on this. The more we stripped out the less useful the snapshots became. Snapshots in general create contextual problems. An author and reviewer can only make sense of a diff if they know the correct design intent. By aligning our snapshots now to the actual figma design via visual regression tests, we do not have to do a translation in our heads. 

Yes, the underlying markup is important. I like @AugustinMauroy's nod to that in the comments, but I think this change will give us more confidence

**Pull request merged**